eth/db/journal.py

```diff
@@ -376,14 +411,17 @@ def diff(self) -> DBDiff:
         visited_keys = set()  # type: Set[bytes]
 
         # Iterate in reverse, so you can skip over any keys from old checkpoints.
-        # This is purely for performance, not correctness.
+        # This is required so that when a key is created and then deleted in the journal,
+        #   we don't add the delete to the diff. (We simply omit the change altogether)
         for changeset in reversed(self.journal.journal_data.values()):
```

Comment:
Ah, thanks!

Interesting, I'm kind of shocked that this didn't trigger a consensus test failure, where storage elements are still available after deletion! They would have to be created just before deletion, but that seems like an obvious thing to check.

I don't understand the suggested solution (it seemed to error for me). But the test case was the important part. I've got it working now :+1:

---

> This is a weak code smell but eth/db/journal.py seems like it's waiting for an abstraction to be found. Starting with `self.journal.journal_data`, it's a little strange that `JournalDB` knows so much about how to read changesets when `Journal` already performs this operation internally.

Yeah, this whole module is a mess. The split of responsibilities doesn't make much sense to me.

Fair criticism that `diff()` is really dealing with `Journal` internals, for no good reason. I moved it inside `Journal`.

---

> One option might be: Instead of `Journal` maintaining an explicit list of changesets it holds onto the head (highest precidence) of a linked list of changesets. Each changeset handles `__getitem__` and `__delitem__` and such by looking first at itself and then recursing into the next changeset. `clear()` is easy to express, it's a subclass which doesn't recurse. Manipulating this stack might be easier that the current logic. For example, `_reapply_changeset_to_journal` could just pass the same stack back to Journal.

Yeah, this seems reasonable to me at first glance. But it's also a tangent of a tangent of a thing I'm trying to get done. As long as we can get it correct for now, I'm okay putting off a refactor for a bit.

Reply:
> I don't understand the suggested solution (it seemed to error for me)

Ah, sorry, should have said the snippet wasn't a solution to the bug, but it might be a step closer to de-duplicating the changeset iteration logic.

> Yeah, this seems reasonable to me at first glance. But it's also a tangent of a tangent of a thing I'm trying to get done. As long as we can get it correct for now, I'm okay putting off a refactor for a bit.

And yeah, totally understood.